src/main/java/com/example/spring/service/cart/CartServiceImpl.java

```diff
@@ -0,0 +1,105 @@
+package com.example.spring.service.cart;
+
+import com.example.spring.dto.cartitem.CartItemRequestDto;
+import com.example.spring.dto.cartitem.UpdateCartItemRequestDto;
+import com.example.spring.dto.shoppingcart.ShoppingCartDto;
+import com.example.spring.exception.EntityNotFoundException;
+import com.example.spring.mapper.CartItemMapper;
+import com.example.spring.mapper.ShoppingCartMapper;
+import com.example.spring.model.Book;
+import com.example.spring.model.CartItem;
+import com.example.spring.model.ShoppingCart;
+import com.example.spring.model.User;
+import com.example.spring.repository.book.BookRepository;
+import com.example.spring.repository.cartitem.CartItemRepository;
+import com.example.spring.repository.shoppingcart.ShoppingCartRepository;
+import jakarta.transaction.Transactional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CartServiceImpl implements CartService {
+    private final CartItemMapper cartItemMapper;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto cartItemRequestDto) {
+        Book book = bookRepository.findById(cartItemRequestDto.bookId())

```

Comment:
Suggestion (look at another implementation)

```
Book book = bookRepository.findById(itemDto.bookId())
                .orElseThrow(() -> new EntityNotFoundException(
                        String.format("Book with id: %d is not found", itemDto.bookId())
                ));
        ShoppingCart cart = repository.findByUserId(userId);
        cart.getCartItems().stream()
                .filter(item -> item.getBook().getId().equals(itemDto.bookId()))
                .findFirst()
                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity() + itemDto.quantity()),
                        () -> addCartItemToCart(itemDto, book, cart));
        repository.save(cart);
        return cartMapper.toDto(cart);
```

Where 

```
private void addCartItemToCart(CartItemRequestDto itemDto, Book book, ShoppingCart cart) {
        CartItem cartItem = itemMapper.toCartItem(itemDto);
        cartItem.setBook(book);
        cart.addItemToCart(cartItem);
    }
```

and in ShoppingCart entity 

```
public void addItemToCart(CartItem cartItem) {
        cartItem.setShoppingCart(this);
        cartItems.add(cartItem);
    }
```

Reply:
good solution, but i like mine more